src/core/shape_inference/include/nms_shape_inference.hpp

```diff
@@ -257,20 +245,17 @@ std::vector<TRShape> shape_infer(const NonMaxSuppression* op,
     const auto& boxes_shape = input_shapes[0];
     const auto& scores_shape = input_shapes[1];
 
-    auto output_shapes = std::vector<TRShape>{TRShape{TDim(-1), 3}};
-    if (boxes_shape.rank().is_static()) {
-        const auto max_out_boxes_per_class = get_input_const_data_as<TRShape, int64_t>(op, 2, ta);
-        const auto max_out_class_boxes = max_out_boxes_per_class ? max_out_boxes_per_class->front() : dim::inf_bound;
-
-        const auto& num_boxes = boxes_shape[1];
-        auto& selected_boxes = output_shapes[0][0];
-        if (num_boxes.is_static()) {
-            selected_boxes = std::min(num_boxes.get_length(), static_cast<V>(max_out_class_boxes));
-        }
-
-        if (scores_shape.rank().is_static()) {
-            selected_boxes *= scores_shape[0].get_max_length();
-            selected_boxes *= scores_shape[1].get_max_length();
+    auto output_shapes = std::vector<TRShape>{TRShape{TDim(dim::inf_bound), 3}};
+
+    if (const auto max_out_boxes_per_class = get_input_const_data_as<TRShape, int64_t>(op, 2, ta)) {
+        if (boxes_shape.rank().is_static() && scores_shape.rank().is_static()) {
+            const auto& num_boxes = boxes_shape[1];
+            if (num_boxes.is_static()) {
+                auto& selected_boxes = output_shapes[0][0];
+                selected_boxes = std::min(num_boxes.get_length(), static_cast<V>(max_out_boxes_per_class->front()));
+                selected_boxes *= scores_shape[0].get_max_length();
+                selected_boxes *= scores_shape[1].get_max_length();
+            }
```

Comment:
In fact `get_input_const_data_as<TRShape, int64_t>(op, 2, ta)` is the more expensive part, 
I would suggest to call as the last check.
```suggestion
    auto output_shapes = std::vector<TRShape>{TRShape{TDim(dim::inf_bound), 3}};

    if (boxes_shape.rank().is_static() && scores_shape.rank().is_static()) {
        if (const auto max_out_boxes_per_class = get_input_const_data_as<TRShape, int64_t>(op, 2, ta)) {
            const auto& num_boxes = boxes_shape[1];
            if (num_boxes.is_static()) {
                auto& selected_boxes = output_shapes[0][0];
                selected_boxes = std::min(num_boxes.get_length(), static_cast<V>(max_out_boxes_per_class->front()));
                selected_boxes *= scores_shape[0].get_max_length();
                selected_boxes *= scores_shape[1].get_max_length();
            }
```